core/src/main/java/com/scalar/db/storage/dynamo/DynamoTableMetadataManager.java

```diff
@@ -127,19 +153,174 @@ private DataType convertDataType(String columnType) {
+      }
+    } catch (DynamoDbException e) {
+      throw new StorageRuntimeException("getting metadata table description failed");
+    }
   }
 
   @Override
   public void addTableMetadata(String namespace, String table, TableMetadata metadata) {
-    // TODO To implement
-    throw new UnsupportedOperationException();
+    createMetadataTableIfNotExist();
+    Map<String, AttributeValue> itemValues = new HashMap<>();
+
+    // Add metadata
+    itemValues.put(
+        TABLE,
+        AttributeValue.builder().s(getFullTableName(namespacePrefix, namespace, table)).build());
+    Map<String, AttributeValue> columns = new HashMap<>();
+    for (String columnName : metadata.getColumnNames()) {
+      columns.put(
+          columnName,
+          AttributeValue.builder()
+              .s(metadata.getColumnDataType(columnName).name().toLowerCase())
+              .build());
+    }
+    itemValues.put(COLUMNS, AttributeValue.builder().m(columns).build());
+    itemValues.put(
+        PARTITION_KEY,
+        AttributeValue.builder()
+            .l(
+                metadata.getPartitionKeyNames().stream()
+                    .map(pKey -> AttributeValue.builder().s(pKey).build())
+                    .collect(Collectors.toList()))
+            .build());
+    itemValues.put(
+        CLUSTERING_KEY,
+        AttributeValue.builder()
+            .l(
+                metadata.getClusteringKeyNames().stream()
+                    .map(pKey -> AttributeValue.builder().s(pKey).build())
+                    .collect(Collectors.toList()))
+            .build());
+    itemValues.put(
+        SECONDARY_INDEX, AttributeValue.builder().ss(metadata.getSecondaryIndexNames()).build());
+
+    PutItemRequest request =
+        PutItemRequest.builder()
+            .tableName(getFullTableName(namespacePrefix, METADATA_NAMESPACE, METADATA_TABLE))
+            .item(itemValues)
+            .build();
+
+    try {
+      client.putItem(request);
+    } catch (DynamoDbException e) {
+      throw new StorageRuntimeException(
+          "adding meta data for table "
+              + getFullTableName(namespacePrefix, namespace, table)
+              + " failed",
+          e);
+    }
   }
 
   @Override
   public Set<String> getTableNames(String namespace) {
-    // TODO To implement
-    throw new UnsupportedOperationException();
+    Set<String> tableSet = new HashSet<>();
+    try {
+      ListTablesResponse listTablesResponse = client.listTables();
+      List<String> tableNames = listTablesResponse.tableNames();
+      for (String tableName : tableNames) {
+        if (tableName.startsWith(getFullNamespaceName(namespacePrefix, namespace))) {
+          tableSet.add(tableName);
+        }
+      }
+    } catch (DynamoDbException e) {
+      throw new StorageRuntimeException("retrieving the table names failed", e);
+    }
+    return tableSet;
+  }
+
+  private void createMetadataTableIfNotExist() throws StorageRuntimeException {
+    if (!metadataTableExists()) {
```

Comment:
We can do `early return` here to reduce nest:
```suggestion
    if (metadataTableExists()) {
      return;
    }
```